boot/freeldr/freeldr/arch/realmode/fathelp.inc

```diff
@@ -158,40 +170,101 @@ GetFatEntry16:
 // On entry AX has cluster number
 // On return AX has FAT entry for that cluster
 GetFatEntry12:
-    push  ax
-    mov   cx,ax
-    shr   ax,1
-    add   ax,cx                     // AX = AX * 1.5 (AX = AX + (AX / 2)) (since FAT12 entries are 12 bits)
-
-    mov   bx, HEX(7000)
-    mov   es,bx
-    mov   bx,ax                     // Put FAT entry offset into BX
-    mov   ax, es:[bx]               // Get FAT entry
+    push  ax                        // Save original cluster number
+    mov   cx, ax
+    shr   ax, 1
+    add   ax, cx                    // AX = AX * 1.5 (AX = AX + (AX / 2)) (since FAT12 entries are 12 bits)
+
+    mov cx, ax                      // Save this for later in CX
+
+    xor dx, dx // Dividend
+    mov bx, word ptr [bp+BytesPerSector]    // FAT Sector Number = original AX / BytesPerSector
+    div bx
+
+    xor dx, dx
+    add ax, word ptr [bp+HiddenSectors]     // FAT Sector Number += HiddenSectors
+    adc dx, word ptr [bp+HiddenSectors+2]
+    add ax, word ptr [bp+ReservedSectors]   // FAT Sector Number += ReservedSectors
+    adc dx, 0
+
+    dec  bx
+    and  cx, bx                     // FAT Offset Within Sector = CX % BytesPerSector
+                                    // DX:AX holds logical FAT sector number
+                                    // CX holds FAT entry offset
+
+    call LoadFatSector              // Get FAT entry in AX
+
     pop   cx                        // Get cluster number from stack
-    and   cx,1
+    and   cx, 1
     jz    UseLow12Bits
     and   ax, HEX(0fff0)
-    shr   ax,4
+    shr   ax, 4
     jmp   GetFatEntry12_Done
 
 UseLow12Bits:
     and   ax, HEX(0fff)
 
 GetFatEntry12_Done:
+    ret
+
+
+/*
+ * This code is directly adapted from fat32.S
+ * BUT: Use DX:AX instead of EAX for the FAT logical sector number.
+ * CX: FAT entry offset.
+ */
+LoadFatSector:
+    push  cx                        // Save FAT entry offset
+
+//
+// FIXME: Why loading the sectors at 7000h do fail?
+// In the meantime, do as with FAT32 and load the sectors
+// at 9000h instead; this seems to work fine...
+//
+    mov   bx, HEX(9000)             // We will load it to [9000:0000h]
+    mov   es, bx
+
+    // DX:AX holds logical FAT sector number
+    // Check if we have already loaded it
+//
+// FIXME: Why using cs:[...] does not work?
+//
+    cmp   ax, word ptr [bp+FatSectorInCache]
+    jne   DoLoad
+    cmp   dx, word ptr [bp+FatSectorInCache+2]
+    je    LoadFatSectorAlreadyLoaded
+
+DoLoad:
+    mov word ptr [bp+FatSectorInCache], ax
+    mov word ptr [bp+FatSectorInCache+2], dx
+    xor bx, bx
+    // Read 2 consecutive sectors, so that, in FAT12, entry data
```

Comment:
If an entry starts on the cache boundary you only neeed a `rep movsw // (move cached sector 2 to the cached sector 1)` and read a single sector to the cache.

Reply:
The FAT entry data being 12 bits, there is the possibility that the evaluated FAT sector offset points to the very last byte of the 1st cached sector, so 8 of the 12 bits will be there, but the last 4 bits will be in the next sector.